src/crimson/osd/pg_backend.cc

```diff
@@ -1120,36 +1130,27 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::getxattr(
   });
 }
 
-PGBackend::get_attr_ierrorator::future<ceph::bufferlist>
-PGBackend::getxattr(
-  const hobject_t& soid,
-  std::string&& key) const
-{
-  return seastar::do_with(key, [this, &soid](auto &key) {
-    return store->get_attr(coll, ghobject_t{soid}, key);
-  });
-}
-
 PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
   const ObjectState& os,
+  const ObjectContext::attr_cache_t& attr_cache,
   OSDOp& osd_op,
   object_stat_sum_t& delta_stats) const
 {
-  return store->get_attrs(coll, ghobject_t{os.oi.soid}).safe_then(
-    [&delta_stats, &osd_op](auto&& attrs) {
-    std::vector<std::pair<std::string, bufferlist>> user_xattrs;
-    ceph::bufferlist bl;
-    for (auto& [key, val] : attrs) {
-      if (key.size() > 1 && key[0] == '_') {
-	bl.append(std::move(val));
-	user_xattrs.emplace_back(key.substr(1), std::move(bl));
-      }
+  if (std::empty(attr_cache)) {
+    return crimson::ct_error::enodata::make();
+  }
+  std::vector<std::pair<std::string, bufferlist>> user_xattrs;
+  ceph::bufferlist bl;
+  for (auto& [key, val] : attr_cache) {
+    if (key.size() > 1 && key[0] == '_') {
+      bl.append(std::move(val));
+      user_xattrs.emplace_back(key.substr(1), std::move(bl));
     }
-    ceph::encode(user_xattrs, osd_op.outdata);
-    delta_stats.num_rd++;
-    delta_stats.num_rd_kb += shift_round_up(bl.length(), 10);
-    return get_attr_errorator::now();
-  });
+  }
+  ceph::encode(user_xattrs, osd_op.outdata);
+  delta_stats.num_rd++;
+  delta_stats.num_rd_kb += shift_round_up(bl.length(), 10);
+  return get_attr_errorator::now();
```

Comment:
Same here, seems `PGBackend::get_xattrs` no longer needs to return a future now that we are not querying the store.